Add NumDescendants to hoeffding tree

sam0410 opened the pull request:
#977

sam0410 commented:
Hi @rcurtin !
Can you please review this.
Thanks!

mlpack-jenkins commented:
Can one of the admins verify this patch?

rcurtin commented:
Can you fix the style please?  https://github.com/mlpack/mlpack/wiki/DesignGuidelines#style-guidelines

Also, could you add another test?  Right now you are only testing that a node with no children has no descendants, but we should test on a larger tree also.

sam0410 commented:
Hi @rcurtin !
As I was going through hoeffding_tree_test.cpp, I noticed that in several places,
```
  arma::mat dataset(3, 9000);
  arma::Row<size_t> labels(9000);
  for (size_t i = 0; i < 9000; i += 3)
  {
    dataset(0, i) = mlpack::math::Random();
    labels[i] = 0;

    dataset(0, i + 1) = mlpack::math::Random();
    dataset(1, i + 1) = mlpack::math::Random() - 1.0;
    dataset(2, i + 1) = mlpack::math::Random() + 0.5;
    labels[i + 1] = 2;

    dataset(0, i + 2) = mlpack::math::Random();
    dataset(1, i + 2) = mlpack::math::Random() + 1.0;
    dataset(2, i + 2) = mlpack::math::Random() + 0.8;
    labels[i + 2] = 1;
  }
```

shouldn't it be

```
  arma::mat dataset(3, 9000);
  arma::Row<size_t> labels(9000);
  for (size_t i = 2; i < 9000; i += 3)
  {
    dataset(0, i) = mlpack::math::Random();
    labels[i] = 0;

    dataset(0, i - 1) = mlpack::math::Random();
    dataset(1, i - 1) = mlpack::math::Random() - 1.0;
    dataset(2, i - 1) = mlpack::math::Random() + 0.5;
    labels[i - 1] = 2;

    dataset(0, i - 2) = mlpack::math::Random();
    dataset(1, i - 2) = mlpack::math::Random() + 1.0;
    dataset(2, i - 2) = mlpack::math::Random() + 0.8;
    labels[i - 2] = 1;
  }
```

rcurtin commented:
I'm not sure I see the functional difference between those two snippets.  This looks just about ready, if you can fix the compilation error I think it could be merged.

zoq commented:
@sam0410 The compiler can't deduce the template parameter ``T`` of ``T MapString(const std::string& string, ...``, you have to specify the type: ``info.MapString<double>("cat0", 0);``.

sam0410 commented:
Hi @zoq !
Can you please help me with the error. I can't figure it out.
error: no matching function for call to ‘std::stack<mlpack::tree::HoeffdingTree<>*>::push(mlpack::tree::HoeffdingTree<mlpack::tree::GiniImpurity, HoeffdingTreeTest::HoeffdingSizeTNumericSplit, mlpack::tree::HoeffdingCategoricalSplit>* const)’
   stack.push(this); // push the current tree

zoq commented:
Here is a patch: https://gist.github.com/zoq/ddaeb93d498b98a487d063c3990a7363 bascially what I did is to use ``HoeffdingTree*`` instead of ``HoeffdingTree<>*``, if you write ``HoeffdingTree<>`` you use the default parameter.

Also the mapping looks strange, I think you should use:

```
info.MapString<size_t>("cat0", 0);
info.MapString<size_t>("cat1", 0);
info.MapString<size_t>("cat2", 0);
info.MapString<size_t>("cat3", 0);
info.MapString<size_t>("cat4", 0);
info.MapString<size_t>("cat5", 0);
info.MapString<size_t>("cat6", 0);
info.MapString<size_t>("cat0", 1);
info.MapString<size_t>("cat1", 1);
info.MapString<size_t>("cat2", 1);
info.MapString<size_t>("cat0", 2);
info.MapString<size_t>("cat1", 2);
```
maybe I missed something?

sam0410 commented:
Hi @zoq !
I changed the function just as you suggested in the patch.  But it said
`mlpack::tree::HoeffdingTree<>::NumDescendants()’ does not match any in class`
 so I changed it to 
`size_t mlpack::tree::HoeffdingTree<FitnessFunction, NumericSplitType,` `CategoricalSplitType>::NumDescendants()`
is this ok?

zoq commented:
> I changed it to
> size_t mlpack::tree::HoeffdingTree<FitnessFunction, NumericSplitType, CategoricalSplitType>::NumDescendants()
> is this ok?

Sure, nice catch, I missed that. Let me know if you need help with the style issues.

sam0410 commented:
Hi @zoq !
the SVDBatchMomentumTest is failing that is why the travis is failing.
Thankyou for your patience.

sam0410 commented:
Thanks @rcurtin ! I'll add my name in my later contributions, you can merge it .

rcurtin commented:
Ok, I will leave this open for at least 3 more days in case anyone else has any comments.  If you have a chance and want to add your name then, that's fine, otherwise it can happen later.

sam0410 commented:
sure @rcurtin  :)

zoq commented on src/mlpack/methods/hoeffding_trees/hoeffding_tree_impl.hpp:
```diff
@@ -500,7 +501,28 @@ size_t HoeffdingTree<
+{
+  size_t nodes = 0;
+  std::stack<HoeffdingTree*> stack;
+  stack.push(this); // push the current tree
```
Minor style issue, can you take a look at: https://github.com/mlpack/mlpack/wiki/DesignGuidelines#comments.
Basically if you do: ``Push the current tree.`` it's fine.

zoq commented on src/mlpack/tests/hoeffding_tree_test.cpp:
```diff
@@ -569,6 +569,73 @@ BOOST_AUTO_TEST_CASE(HoeffdingTreeSimpleDatasetTest)
+  for (size_t i = 0; i < 500; ++i)
+    streamTree.Train(dataset.col(i), labels[i]);
+
+  BOOST_REQUIRE_EQUAL(streamTree.NumDescendants(), 0); // one label
```
Pedantic style issue, please use complete sentences, also you can put the comment above ``BOOST_REQUIRE_EQUAL`` if you run into the 80 columns length limit.

zoq commented on src/mlpack/tests/hoeffding_tree_test.cpp:
```diff
@@ -569,6 +569,73 @@ BOOST_AUTO_TEST_CASE(HoeffdingTreeSimpleDatasetTest)
+    labels(i - 2) = 2;
+  }
+
+  // Now train the streaming decision tree
```
Another pedantic style issue missing ``.`` at the end.

rcurtin commented:
Thanks!  I am glad to have this one done with.  I added some comments and made NumDescendants() const in 4d102b7.  If you'd like to open another PR adding your name and email to `COPYRIGHT.txt` and `src/mlpack/core.hpp`, I'll go ahead and merge it.

sam0410 commented:
Thanks @rcurtin ! I added a PR for that
